build_tools/scripts/export_submodule_head.py

```diff
@@ -0,0 +1,225 @@
+#!/usr/bin/env python
+"""Exports a submodule head to a branch in one of our fork repos.
+
+If patches have been applied to a local submodule for which we
+host a fork repository, this script will "export" those patches
+to specially maintained branches in the fork such that the
+head commit is available indefinitely and can be used safely
+as a submodule commit.
+
+This will create branches in the fork named like:
+  sm-iree-{branch}
+
+(i.e. all historical patch commits for `main` will be on
+sm-iree-main or if they derive from a specific integrate or
+branch on sm-iree-integrate_....)
+
+The forks have branch protection enabled for branches starting
+with "sm-" so that they cannot be force pushed.
+
+This is all accomplished by creating special merge commits and
+associating trees with them explicitly. Most folks don't need
+to understand that: they just need to know that if you run
+this script, the current commit at your submodule head will
+be available "forever".
+
+Usage:
+  ./scripts/export_submodule_head.py third_party/llvm-project
+  ./scripts/export_submodule_head.py third_party/stablehlo
+  ./scripts/export_submodule_head.py third_party/torch-mlir
+"""
+from typing import Optional
+
+import argparse
+from pathlib import Path
+import re
+import shlex
+import subprocess
+import sys
+
+
+def export_submodule_head(args, submodule_rel_path: str):
+    super_repo_path = args.repo_path
+    super_repo_name = args.repo_path.name
+    super_branch = git_current_branch(repo_dir=args.repo_path)
+    print(f"Super-repo '{super_repo_name}' is on branch '{super_branch}'")
+    submodule_path = super_repo_path / submodule_rel_path
+    print(f"Operating on submodule {submodule_path}")
+    if not check_origin_update_help(submodule_path):
+        return 1
+    git_fetch(repository="origin", repo_dir=submodule_path)
```

Comment:
Can you make "origin" a flag variable instead of an inlined string? I use "upstream", for example:
```
D:\dev\projects\iree (main -> upstream)
λ git remote -v
scotttodd       https://github.com/ScottTodd/iree.git (fetch)
scotttodd       https://github.com/ScottTodd/iree.git (push)
upstream        https://github.com/openxla/iree (fetch)
upstream        https://github.com/openxla/iree (push)
```

We have a few bash scripts in the repo that use 'upstream' by default also: https://github.com/openxla/iree/blob/ef35fa2dc8fad6724b7608a074e3c4c0d0550ae1/build_tools/scripts/git/git_update.sh#L9-L20

Fine to default to "origin", I don't think the other git scripts are used by enough people to really push for standardized remote names.

Reply:
Oh right, the submodule remotes. Yeah:

```
D:\dev\projects\iree\third_party\llvm-project (HEAD detached at 419c919)
λ git remote -v
origin  https://github.com/shark-infra/llvm-project.git (fetch)
origin  https://github.com/shark-infra/llvm-project.git (push)
scotttodd       https://github.com/ScottTodd/llvm-project.git (fetch)
scotttodd       https://github.com/ScottTodd/llvm-project.git (push)
upstream        https://github.com/llvm/llvm-project.git (fetch)
upstream        https://github.com/llvm/llvm-project.git (push)
```